requests_kerberos/kerberos_.py

```diff
@@ -86,70 +84,71 @@ class HTTPKerberosAuth(AuthBase):
 
-    def generate_request_header(self, response):
+    def generate_request_header(self, response, host, is_preemptive=False):
         """
         Generates the GSSAPI authentication token with kerberos.
 
-        If any GSSAPI step fails, return None.
+        If any GSSAPI step fails, raise KerberosExchangeError
+        with failure detail.
 
         """
-        host = urlparse(response.url).hostname
 
         # Flags used by kerberos module.
         gssflags = kerberos.GSS_C_MUTUAL_FLAG | kerberos.GSS_C_SEQUENCE_FLAG
         if self.delegate:
             gssflags |= kerberos.GSS_C_DELEG_FLAG
 
         try:
+            kerb_stage = "authGSSClientInit()"
             result, self.context[host] = kerberos.authGSSClientInit(
                 "{0}@{1}".format(self.service, host), gssflags=gssflags)
-        except kerberos.GSSError as error:
-            log.error("generate_request_header(): authGSSClientInit() failed:")
-            log.exception(error)
-            return None
 
-        if result < 1:
-            log.error("generate_request_header(): authGSSClientInit() failed: "
-                      "{0}".format(result))
-            return None
+            if result < 1:
+                raise EnvironmentError(result, kerb_stage)
 
-        try:
+            # if we have a previous response from the server, use it to continue
+            # the auth process, otherwise use an empty value
+            negotiate_resp_value = '' if is_preemptive else _negotiate_value(response)
+
+            kerb_stage = "authGSSClientStep()"
             result = kerberos.authGSSClientStep(self.context[host],
-                                                _negotiate_value(response))
-        except kerberos.GSSError as error:
-            log.exception(
-                "generate_request_header(): authGSSClientStep() failed:")
-            log.exception(error)
-            return None
+                                                negotiate_resp_value)
 
-        if result < 0:
-            log.error(
-                "generate_request_header(): authGSSClientStep() failed: "
-                "{0}".format(result))
-            return None
+            if result < 0:
+                raise EnvironmentError(result, kerb_stage)
 
-        try:
+            kerb_stage = "authGSSClientResponse()"
             gss_response = kerberos.authGSSClientResponse(self.context[host])
+
+            return "Negotiate {0}".format(gss_response)
+
         except kerberos.GSSError as error:
             log.exception(
-                "generate_request_header(): authGSSClientResponse() failed:")
+                "generate_request_header(): {0} failed:".format(kerb_stage))
             log.exception(error)
-            return None
+            raise KerberosExchangeError("%s failed: %s" % (kerb_stage, str(error.args)))
 
-        return "Negotiate {0}".format(gss_response)
+        except EnvironmentError as error:
```

Comment:
This was another, "worked as written but not as intended"- had intended to ensure the EnvironmentException matched the return code, then translate it to KerberosExchangeError if so (or re-raise if not). Added that (though we could just as easily use a custom internal error type)- the  "translate return codes into an exception" stuff is always messy. Updated.